server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,443 @@
+             * We will be reading them directory from the local directory
+             */
+            if (isTempFile(name)) {
+                return localDirectory.openInput(name, context);
+            }
+            /**
+             * Return directly from the FileCache (via TransferManager) if complete file is present
+             */
+            else if (fileCache.get(localDirectory.getDirectory().resolve(name)) != null) {
+                logger.trace("Complete file found in FileCache");
+                BlobFetchRequest blobFetchRequest = BlobFetchRequest.builder()
+                    .directory(localDirectory)
+                    .fileName(name)
+                    // position and length are not required here since this is a complete file, just adding dummy values for validation
+                    .blobParts(new ArrayList<>(Arrays.asList(new BlobFetchRequest.BlobPart(name, 0, 1))))
+                    .build();
+                return transferManager.fetchBlob(blobFetchRequest);
+            }
+            /**
+             * If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via OnDemandCompositeBlockIndexInput
+             */
+            else {
+                logger.trace("Complete file not in FileCache, to be fetched in Blocks from Remote");
+                RemoteSegmentMetadata remoteSegmentMetadata = remoteDirectory.readLatestMetadataFile();
+                RemoteSegmentStoreDirectory.UploadedSegmentMetadata uploadedSegmentMetadata = remoteSegmentMetadata.getMetadata().get(name);
+                /**
+                 * TODO : Refactor FileInfo and OnDemandBlockSnapshotIndexInput to more generic names as they are not Remote Snapshot specific
+                 */
+                BlobStoreIndexShardSnapshot.FileInfo fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
+                    name,
+                    new StoreFileMetadata(name, uploadedSegmentMetadata.getLength(),
+                        uploadedSegmentMetadata.getChecksum(), Version.LATEST),
+                    null
+                );
+                return new OnDemandBlockSnapshotIndexInput(fileInfo, localDirectory, transferManager);
+                //return new OnDemandCompositeBlockIndexInput(remoteDirectory, name, localDirectory, fileCache, context);
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Acquires and returns a {@link Lock} for a file with the given name.
+     * @param name the name of the lock file
+     * @throws LockObtainFailedException (optional specific exception) if the lock could not be
+     * obtained because it is currently held elsewhere.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public Lock obtainLock(String name) throws IOException {
+        logger.trace("obtainLock() called {}", name);
+        writeLock.lock();
```

Comment:
am curious why  all this locking necessary in this class?  I would expect the engine to control write locks where necessary?

Reply:
Based on the offline discussion, the locking was introduced to  avoid certain race conditions with delete & listAll. I think we should cleanup the locks(in this PR or as follow up). 

> For example: deleteFile should delete just local copy or both local and remote copy of a file ? Similarly close should close both local or remote directory here ? Will be good to capture these interactions via test and simulating the behavior on what is expected and also adding proper java docs for these APIs

Agreed. We need to clearly document the expected behavior in such cases.